tests/preconditions_tests.cpp

```diff
@@ -0,0 +1,237 @@
+      0x6e, 0x63, 0x5f, 0x65, 0x00, 0x03, 0x03, 0x0a, 0x09, 0x02, 0x02, 0x02,
+      0x02, 0x02, 0x02, 0x02, 0x02, 0x02, 0x05, 0x03, 0x01, 0x00, 0x01, 0x07,
+      0x7f, 0x09, 0x0b, 0x63, 0x61, 0x6c, 0x6c, 0x5f, 0x74, 0x65, 0x73, 0x74,
+      0x5f, 0x61, 0x00, 0x05, 0x0b, 0x63, 0x61, 0x6c, 0x6c, 0x5f, 0x74, 0x65,
+      0x73, 0x74, 0x5f, 0x62, 0x00, 0x06, 0x0b, 0x63, 0x61, 0x6c, 0x6c, 0x5f,
+      0x74, 0x65, 0x73, 0x74, 0x5f, 0x63, 0x00, 0x07, 0x0b, 0x63, 0x61, 0x6c,
+      0x6c, 0x5f, 0x74, 0x65, 0x73, 0x74, 0x5f, 0x64, 0x00, 0x08, 0x0b, 0x63,
+      0x61, 0x6c, 0x6c, 0x5f, 0x74, 0x65, 0x73, 0x74, 0x5f, 0x65, 0x00, 0x09,
+      0x0b, 0x63, 0x61, 0x6c, 0x6c, 0x5f, 0x74, 0x65, 0x73, 0x74, 0x5f, 0x66,
+      0x00, 0x0a, 0x0b, 0x63, 0x61, 0x6c, 0x6c, 0x5f, 0x74, 0x65, 0x73, 0x74,
+      0x5f, 0x67, 0x00, 0x0b, 0x0b, 0x63, 0x61, 0x6c, 0x6c, 0x5f, 0x74, 0x65,
+      0x73, 0x74, 0x5f, 0x68, 0x00, 0x0c, 0x0b, 0x63, 0x61, 0x6c, 0x6c, 0x5f,
+      0x74, 0x65, 0x73, 0x74, 0x5f, 0x6a, 0x00, 0x0d, 0x0a, 0x8b, 0x01, 0x09,
+      0x0b, 0x00, 0x41, 0x2a, 0x43, 0x14, 0xae, 0x29, 0x42, 0x10, 0x01, 0x0b,
+      0x10, 0x00, 0x41, 0x90, 0xc0, 0x00, 0x41, 0xa8, 0xc0, 0x00, 0x41, 0xa8,
+      0xc0, 0x00, 0x10, 0x00, 0x0b, 0x0b, 0x00, 0x41, 0x20, 0x43, 0x00, 0x00,
+      0x28, 0x42, 0x10, 0x01, 0x0b, 0x0b, 0x00, 0x41, 0x2a, 0x43, 0x00, 0x00,
+      0x00, 0x42, 0x10, 0x01, 0x0b, 0x0a, 0x00, 0x41, 0x00, 0x41, 0x00, 0x41,
+      0x00, 0x10, 0x00, 0x0b, 0x16, 0x00, 0x43, 0x14, 0xae, 0x29, 0x42, 0x41,
+      0x90, 0xc0, 0x00, 0x41, 0x90, 0xc0, 0x00, 0x41, 0x07, 0x41, 0xcd, 0x00,
+      0x10, 0x02, 0x0b, 0x16, 0x00, 0x43, 0x14, 0xae, 0x29, 0x42, 0x41, 0xa8,
+      0xc0, 0x00, 0x41, 0xa8, 0xc0, 0x00, 0x41, 0x04, 0x41, 0xcd, 0x00, 0x10,
+      0x02, 0x0b, 0x04, 0x00, 0x10, 0x03, 0x0b, 0x16, 0x00, 0x43, 0x14, 0xae,
+      0x29, 0x42, 0x41, 0xa8, 0xc0, 0x00, 0x41, 0xa8, 0xc0, 0x00, 0x41, 0x04,
+      0x41, 0xcd, 0x00, 0x10, 0x04, 0x0b, 0x0b, 0x1e, 0x02, 0x00, 0x41, 0x90,
+      0xc0, 0x00, 0x0b, 0x0b, 0x68, 0x65, 0x6c, 0x6c, 0x6f, 0x68, 0x65, 0x6c,
+      0x6c, 0x6f, 0x00, 0x00, 0x41, 0xa8, 0xc0, 0x00, 0x0b, 0x04, 0xff, 0xff,
+      0xff, 0xff
+   };
+
+   rhf_t::add<&test_func_a, int_equals_42, float_equals_42>("env", "test_func_a");
+   rhf_t::add<&test_func_b, str_equals_hello>("env", "test_func_b");
+   rhf_t::add<&test_func_c, float_equals_42, str_equals_hello, span_and_check>("env", "test_func_c");
+   rhf_t::add<&test_func_d, check2_once>("env", "test_func_d");
+   rhf_t::add<&test_func_e, check2_all>("env", "test_func_e");
+
+   wasm_allocator wa;
+   backend_t bkend(test_wasm, &wa);
+
+   check  = 0;
+   check2 = 0;
+   bkend("env", "call_test_a");
+   bkend("env", "call_test_b");
+   CHECK_MSG("arg != 42", bkend("env", "call_test_c"));
+   CHECK_MSG("arg != 42.42f", bkend("env", "call_test_d"));
+   bkend("env", "call_test_e");
+   CHECK(check == 0);
+   bkend("env", "call_test_f");
+   CHECK(check == 77);
+
+   // reset check
```

Comment:
`check` is not reset despite comment. Or maybe the comment is misplaced?

Reply:
I'm not quite sure what this is intended to test.

Reply:
Hard to say. I think the `check = 0;` below actually was meant to be moved up here.

Reply:
In other words, I think this block of code:
```cpp
   bkend("env", "call_test_f");
   CHECK(check == 77);

   // reset check
   bkend("env", "call_test_f");
   CHECK(check == 77);

   // reset check
   check = 0;
   CHECK_MSG("str != hello", bkend("env", "call_test_g"));
```

makes more sense as:
```cpp
   bkend("env", "call_test_f");
   CHECK(check == 77);

   // reset check
   check = 0;
   bkend("env", "call_test_f");
   CHECK(check == 77);

   CHECK_MSG("str != hello", bkend("env", "call_test_g"));
```

Reply:
Or just:
```cpp
   bkend("env", "call_test_f");
   CHECK(check == 77);

   CHECK_MSG("str != hello", bkend("env", "call_test_g"));
```

and remove the redundant check.